### ISSUE TITLE
Multithreading issue with vcrpy - Inconsistent recording of requests

### ISSUE THREAD

**Issue by stanBienaives** [kevin1024/vcrpy#849]
# Description:

I am encountering an issue with vcrpy when using it in a multithreaded context. The cassette file does not record all the requests when using ThreadPoolExecutor with more than one worker. The issue is not present when using single-threaded execution.

# Test Code:

```python
import vcr  # type: ignore
from concurrent.futures import ThreadPoolExecutor
import requests

vcr = vcr.VCR(record_mode='once', filter_headers=['authorization'], ignore_hosts=['api.smith.langchain.com'])

def test_multithreading():
    with vcr.use_cassette('./test_multithread.yaml', match_on=['method', 'scheme', 'host', 'port', 'path', 'query', 'body']):
        def get_google(num):
            return requests.get(f"https://www.google.com?q={num}")
        with ThreadPoolExecutor(max_workers=2) as executor:
            responses = executor.map(get_google, range(2))

    with open('./test_multithread.yaml', 'r') as f:
        data = f.read()
        assert data.count('https://www.google.com') == 2

def test_singlethreading():
    with vcr.use_cassette('./test_singlethread.yaml', match_on=['method', 'scheme', 'host', 'port', 'path', 'query', 'body']):
        def get_google(num):
            return requests.get(f"https://www.google.com?q={num}")
        with ThreadPoolExecutor(max_workers=1) as executor:
            responses = executor.map(get_google, range(2))

    with open('./test_singlethread.yaml', 'r') as f:
        data = f.read()
        assert data.count('https://www.google.com') == 2
```

# Observed Behavior:

**test_singlethreading** passes as expected, with the cassette recording both requests.
**test_multithreading** fails because the cassette file does not record both requests, resulting in a mismatch.

# Environment:
```
Python 3.12.3
vcrpy==6.0.1
requests==2.31.0
```
# Steps to Reproduce:

Run the provided test code.
Observe that test_multithreading fails while test_singlethreading passes.

# Expected Behavior:

Both test_multithreading and test_singlethreading should pass, with the cassette files correctly recording all the requests made during the tests.

**Comment by stanBienaives** [comment]
interesting enough: 

Adding a delay makes the test pass:
```python
def test_multithreading_record_with_delay():
    with vcr.use_cassette('./test_multithread_with_delay.yaml', match_on=['method', 'scheme', 'host', 'port', 'path', 'query', 'body']):
        def get_google(num):
            time.sleep(num)
            return requests.get(f"https://www.google.com?q={num}")
        with ThreadPoolExecutor(max_workers=2) as executor:
            responses  = executor.map(get_google, range(2))

    with open('./test_multithread_with_delay.yaml', 'r') as f:
        data = f.read()
        assert data.count('https://www.google.com') == 2
```

**Comment by kevin1024** [comment]
@simon-weber I think that’s exactly what’s happening.

I guess we could put a mutex around the unpatch but you would lose some of the benefit of multithreading to begin with.

**Comment by simon-weber** [comment]
I'm not sure there's an easy locking fix inside vcrpy. Locking in force_reset doesn't work since it's not called if patches aren't applied.

What about if patches were left in place but dynamically disabled with a threadlocal? [Wrapt has a way to do this](https://github.com/GrahamDumpleton/wrapt/blob/5c0997c0b5be36da5f621c3473c95f9efe14891e/docs/decorators.rst#dynamically-disabling-decorators) and seems like it could pretty easily replace mock.patch.object.

**Comment by kevin1024** [comment]
Oh good point. Yes, I think that strategy could work.

**Comment by abriere-coveo** [comment]
This is still an issue in the latest version of vcrpy @stanBienaives 

Running the provided test code above, the multithreading test is only recording one of the two HTTP request.

Environment:
```
Python 3.10.18
vcrpy==8.1.0
requests==2.32.4
```

**Comment by abriere-coveo** [comment]
@kevin1024 & @stanBienaives 
Could we re-open this issue since the problem is still ongoing?

Do you prefer we open a new issue?

**Comment by stanBienaives** [comment]
@abriere-coveo Thanks for the follow-up. I am not an active maintainer of this project, so I cannot really advise on the best next step.
I am no longer using vcrpy in my projects — partly because of this concurrency issue, and partly because it significantly slowed down our CI (we used it for end-to-end agent workflow tests).
That said, I still have a few use cases in mind and will very likely come back to it in the future.

**Comment by abriere-coveo** [comment]
@kevin1024 Any chances you could look at this issue?

**Comment by CharString** [comment]
@abriere-coveo Please have a look at https://vcrpy.readthedocs.io/en/latest/contributing.html#collaborators. Any chances you could volunteer for any of the "Needs support" positions?

**Comment by tmiquel** [comment]
I have had a similar occurence in a record mode ONCE where calls to `ignored_hosts` were unpatching the stub libs (thru `force_reset`) while calls that should have matched stored requests were being made (as the stub libs are temporarely unpatched by `force_reset`), while they should get their responses from the cassette.

I applied a local patch to the lib by enforcing a re-entrant thread locks on the applicable functions and it fixes my issue.

Would this kind of fix be acceptable to maintainers?

<img width="2372" height="944" alt="Image" src="https://github.com/user-attachments/assets/7cfcd863-1325-41ff-b72f-522e5013852f" />

**Comment by kevin1024** [comment]
Yes. Though I’m not sure how this will interact with our other adapters (for example, aiohttp)

I would also want to see a pretty comprehensive suite of tests to prove we are thread-safe.
